doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,94 @@
+
+- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
+- Limiting it to only `TIMESTAMP` because
+    - `OFFSET` is already default. We can make it explicit though if we want.
+    - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause.
+    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
+- Limiting to only ascending order because no users have expressed a need for descending.
+
+# Explanation
+
+## Proposal
+
+Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifier should refer to the included TIMESTAMP metadata column.
+
+Examples of valid syntax:
+- `CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY TIMESTAMP )`
+- `CREATE SOURCE ... INCLUDE TIMESTAMP AS ts ENVELOPE UPSERT ( ORDER BY ts )`
+
+Examples of invalid syntax:
+- `CREATE SOURCE ... INCLUDE OFFSET ENVELOPE UPSERT ( ORDER BY TIMESTAMP )` (TIMESTAMP column is not included)
+- `CREATE SOURCE ... INCLUDE PARTITION AS p ENVELOPE UPSERT ( ORDER BY p )` (ORDER BY identifier does not refer to the TIMESTAMP metadata column)
+
+# Reference Explanation
+For envelope upsert, the included metadata columns are appended to the value and eventually persisted. For this feature, we will keep track of the index of the order by field and read the corresponding value to compare when upsert-ing.
+
+Here's an example of the upsert behavior where the source consists of a key, value and the timestamp from metadata and is ordered by the timestamp.
+
+```
+ Previously persisted data        Incoming kafka stream decoded
++------------------------+         +-------------------------+
+| Ok (key1, value1, 100) |         |  Ok (key1, value3, 200) |
+| Ok (key2, value2, 201) |         |  Err(key3, some_error2) |
+| Err(key3, some_error1) |         |  Ok (key2, value4, 200) |
++------------------+-----+         +-----+-------------------+
+                   |                     |
+                   |                     |
+                   |                     |
+                   |                     |
+                   v    Upsert Output    v
+                  +------------------------+
+                  | Ok (key1, value3, 200) |
+                  | Ok (key2, value2, 201) |
+                  | Err(key3, some_error2) |
+                  +------------------------+
+```
+Note: As shown in the example above, the errors are still implicitly ordered by offset as we do not persist any extra metadata for them separately. A later error with the same key will always overwrite a previous one.
```

Comment:
This is surprising, IMO! Perhaps we _should_ be attaching the timestamp column to the errors, too? I don't actually understand in what case we can end up with `Err((key, error_text))`—if Avro decoding of the value fails, but not the key?

Reply:
> if Avro decoding of the value fails, but not the key?
yes!

For some context, a somewhat recent change that made errors retractable: https://github.com/MaterializeInc/materialize/pull/14209. Which also made how errors work very complex.

Errors in the upsert source are a tricky subject. In general, the current "design" (evolved through quite some iterations) assumes that you only `ORDER BY` the offset. That's, for example, why handling errors works the way it does, above: when a new error comes in, it _can_ only have an offset that is larger than the error that we re-hydrated from persist, so it always takes precedence. Ordering by a custom column breaks that.

To fix that, we'd have to graft the `ORDER BY` field on to errors, which feel brittle enough as they are. :see_no_evil: 

Galaxy-brain idea, where I'm not at all sure we should do it. We could have a new implementation of the UPSERT machinery that is only used for new sources when a custom `ORDER BY` is used. We could fix our issues around errors and other things.

Reply:
To clarify further, if say an Error is persisted, we sort of lose the kafka metadata, since those don't get persisted. But for the incoming kafka stream, we will actually have the metadata, so we will have some order value. So, it's more like a comparison between `None` (when hydrating state from persist) with `Some(ts)` (value coming from kafka) and the latter taking precedence.

Reply:
> We could have a new implementation of the UPSERT machinery that is only used for new sources when a custom ORDER BY is used.

Does it mean that the upsert operator will only have the kafka ingestion as the input and there will be no read from persist? 

Actually limiting it like that **would** simplify things 😬.